datasets/flwr_datasets/partitioner/image_semantic_partitioner.py

```diff
@@ -0,0 +1,505 @@
+
+    # pylint: disable=C0415, R0915
+    def _determine_partition_id_to_indices_if_needed(self) -> None:
+        """Create an assignment of indices to the partition indices."""
+        if self._partition_id_to_indices_determined:
+            return
+        try:
+            import torch
+            from scipy.optimize import linear_sum_assignment
+            from sklearn.decomposition import PCA
+            from sklearn.mixture import GaussianMixture
+            from sklearn.preprocessing import StandardScaler
+            from torch.distributions import MultivariateNormal, kl_divergence
+            from torchvision import models
+        except ImportError:
+            raise ImportError(
+                "ImageSemanticPartitioner requires scikit-learn, torch, "
+                "torchvision, scipy, and numpy."
+            ) from None
+        efficient_nets_dict = [
+            (models.efficientnet_b0, models.EfficientNet_B0_Weights.DEFAULT),
+            (models.efficientnet_b1, models.EfficientNet_B1_Weights.DEFAULT),
+            (models.efficientnet_b2, models.EfficientNet_B2_Weights.DEFAULT),
+            (models.efficientnet_b3, models.EfficientNet_B3_Weights.DEFAULT),
+            (models.efficientnet_b4, models.EfficientNet_B4_Weights.DEFAULT),
+            (models.efficientnet_b5, models.EfficientNet_B5_Weights.DEFAULT),
+            (models.efficientnet_b6, models.EfficientNet_B6_Weights.DEFAULT),
+            (models.efficientnet_b7, models.EfficientNet_B7_Weights.DEFAULT),
+        ]
+        backbone = efficient_nets_dict[self._efficient_net_type][0]
+        pretrained_weight = efficient_nets_dict[self._efficient_net_type][1]
+        efficient_net: models.EfficientNet = backbone(weights=pretrained_weight)
+        efficient_net.classifier = torch.nn.Flatten()
+
+        device = torch.device("cpu")
+        if self._use_cuda:
+            if torch.cuda.is_available():
+                device = torch.device("cuda")
+            else:
+                warnings.warn(
+                    "No detected CUDA device, the device fallbacks to CPU.",
+                    UserWarning,
+                    stacklevel=1,
+                )
+        efficient_net.to(device)
+        efficient_net.eval()
+
+        # Generate information needed for Image semantic partitioning
+        self._unique_classes = self.dataset.unique(self._partition_by)
+        assert self._unique_classes is not None
+
+        # Use EfficientNet to extract embeddings
+        images = self._preprocess_dataset_images()
+        embedding_list = []
+        with torch.no_grad():
+            for i in range(0, images.shape[0], self._batch_size):
+                idxs = list(range(i, min(i + self._batch_size, images.shape[0])))
+                batch = torch.tensor(images[idxs], dtype=torch.float, device=device)
+                if batch.shape[1] == 1:
+                    batch = batch.broadcast_to((batch.shape[0], 3, *batch.shape[2:]))
```

Comment:
Let's add a comment here that it's for the purpose of changing the grayscale images to meet the RGB shape.